File: testgen/ui/views/dialogs/column_history_dialog.py
import streamlit as st
from sqlalchemy.sql.expression import func

from testgen.common.models import with_database_session
from testgen.common.models.profiling_run import ProfilingRun
from testgen.ui.components import widgets as testgen
from testgen.ui.components.widgets import testgen_component
from testgen.ui.queries.profiling_queries import COLUMN_PROFILING_FIELDS
from testgen.ui.services.database_service import fetch_one_from_db
from testgen.utils import make_json_safe


def column_history_dialog(*args) -> None:
    st.session_state["column_history_dialog:run_id"] = None
    _column_history_dialog(*args)
    

@st.dialog(title="Column History")
@with_database_session
def _column_history_dialog(
    table_group_id: str,
    schema_name: str,
    table_name: str,
    column_name: str,
    add_date: int,
) -> None:
    testgen.css_class("l-dialog")
    caption_column, loading_column = st.columns([ 0.8, 0.2 ], vertical_alignment="bottom")

    with caption_column:
        testgen.caption(f"Table > Column: <b>{table_name} > {column_name}</b>")

    with loading_column:
        with st.spinner("Loading data ..."):
            profiling_runs = ProfilingRun.select_minimal_where(
                ProfilingRun.table_groups_id == table_group_id,
                ProfilingRun.profiling_starttime >= func.to_timestamp(add_date),
            )
            profiling_runs = [run.to_dict(json_safe=True) for run in profiling_runs]
            run_id = st.session_state.get("column_history_dialog:run_id") or profiling_runs[0]["id"]
            selected_item = get_run_column(run_id, schema_name, table_name, column_name)

    testgen_component(
        "column_profiling_history",
        props={
            "profiling_runs": [
                {
                    "run_id": run["id"],
                    "run_date": run["profiling_starttime"],
                } for run in profiling_runs
            ],
            "selected_item": make_json_safe(selected_item),
        },
        on_change_handlers={
            "RunSelected": on_run_selected,
        }
    )


def on_run_selected(run_id: str) -> None:
    st.session_state["column_history_dialog:run_id"] = run_id


@st.cache_data(show_spinner=False)
def get_run_column(run_id: str, schema_name: str, table_name: str, column_name: str) -> dict:
    query = f"""
    SELECT
        profile_run_id::VARCHAR,
        general_type,
        {COLUMN_PROFILING_FIELDS}
    FROM profile_results
    WHERE profile_run_id = :run_id
        AND schema_name = :schema_name
        AND table_name = :table_name
        AND column_name = :column_name;
    """
    params = {
        "run_id": run_id,
        "schema_name": schema_name,
        "table_name": table_name,
        "column_name": column_name,
    }
    result = fetch_one_from_db(query, params)
    return dict(result) if result else None

File: testgen/ui/components/frontend/js/shared/column_history_dialog.js
import van from '/app/static/js/van.min.js';
import { getValue } from '/app/static/js/utils.js';
import { Dialog } from '/app/static/js/components/dialog.js';
import { ColumnProfilingHistory } from '../data_profiling/column_profiling_history.js';

/**
 * Shared dialog for displaying column profiling history.
 *
 * @param {object} props
 * @param {object} props.historyData - reactive state: set to data object to open, null to close
 *   Shape: { table_name, column_name, profiling_runs, selected_item }
 * @param {function} props.onClose - called when dialog is closed
 * @param {function} props.onRunSelected - called when a profiling run is selected
 */
const ColumnHistoryDialog = (props) => {
    const open = van.state(false);
    const data = van.state(null);

    van.derive(() => {
        const raw = getValue(props.historyData) ?? null;
        data.val = raw;
        open.val = !!raw;
    });

    const onClose = () => {
        open.val = false;
        props.onClose?.();
    };

    const profilingRuns = van.derive(() => data.val?.profiling_runs ?? []);
    const selectedItem = van.derive(() => data.val?.selected_item ?? null);
    const title = van.derive(() => {
        const d = data.val;
        return d ? `Column History: ${d.table_name} > ${d.column_name}` : 'Column History';
    });

    return Dialog(
        { title, open, onClose, width: '60rem' },
        () => data.val
            ? ColumnProfilingHistory({
                profiling_runs: profilingRuns,
                selected_item: selectedItem,
                onRunSelected: props.onRunSelected,
            })
            : '',
    );
};

export { ColumnHistoryDialog };
